### ISSUE TITLE
ABI changes in 1.0.0  - time for a soname bump?

### ISSUE THREAD

**Issue by opoplawski** [Cisco-Talos/clamav#775]
Describe the bug
----------------
I'm comparing libclamav from 0.103.7 (current Fedora) to 1.0.0-rc2 and it's reporting a number of changes to public symbols.  Below are removed functions marked CLAMAV_PUBLIC.
```
$ abipkgdiff  --d1 clamav-lib-debuginfo-0.103.7-4.fc38.x86_64.rpm --d2 clamav-lib-debuginfo-1.0.0-rc2.x86_64.rpm --devel1 clamav-devel-0.103.7-4.fc38.x86_64.rpm --devel2 clamav-devel-1.0.0-rc2.x86_64.rpm clamav-lib-0.103.7-4.fc38.x86_64.rpm clamav-lib-1.0.0-rc2.x86_64.rpm | grep PUBLIC
    [D] 'function void cl_always_gen_section_hash()'    {cl_always_gen_section_hash@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_countsigs(const char*, unsigned int, unsigned int*)'    {cl_countsigs@@CLAMAV_PUBLIC}
    [D] 'function void cl_cvdfree(cl_cvd*)'    {cl_cvdfree@@CLAMAV_PUBLIC}
    [D] 'function cl_cvd* cl_cvdhead(const char*)'    {cl_cvdhead@@CLAMAV_PUBLIC}
    [D] 'function cl_cvd* cl_cvdparse(const char*)'    {cl_cvdparse@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_cvdverify(const char*)'    {cl_cvdverify@@CLAMAV_PUBLIC}
    [D] 'function void cl_debug()'    {cl_debug@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_addref(cl_engine*)'    {cl_engine_addref@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_compile(cl_engine*)'    {cl_engine_compile@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_free(cl_engine*)'    {cl_engine_free@@CLAMAV_PUBLIC}
    [D] 'function long long int cl_engine_get_num(const cl_engine*, cl_engine_field, int*)'    {cl_engine_get_num@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_engine_get_str(const cl_engine*, cl_engine_field, int*)'    {cl_engine_get_str@@CLAMAV_PUBLIC}
    [D] 'function cl_engine* cl_engine_new()'    {cl_engine_new@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_file_props(cl_engine*, clcb_file_props)'    {cl_engine_set_clcb_file_props@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_hash(cl_engine*, clcb_hash)'    {cl_engine_set_clcb_hash@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_meta(cl_engine*, clcb_meta)'    {cl_engine_set_clcb_meta@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_post_scan(cl_engine*, clcb_post_scan)'    {cl_engine_set_clcb_post_scan@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_pre_cache(cl_engine*, clcb_pre_cache)'    {cl_engine_set_clcb_pre_cache@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_pre_scan(cl_engine*, clcb_pre_scan)'    {cl_engine_set_clcb_pre_scan@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_sigload(cl_engine*, clcb_sigload, void*)'    {cl_engine_set_clcb_sigload@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_add_sample(cl_engine*, clcb_stats_add_sample)'    {cl_engine_set_clcb_stats_add_sample@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_decrement_count(cl_engine*, clcb_stats_decrement_count)'    {cl_engine_set_clcb_stats_decrement_count@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_flush(cl_engine*, clcb_stats_flush)'    {cl_engine_set_clcb_stats_flush@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_get_hostid(cl_engine*, clcb_stats_get_hostid)'    {cl_engine_set_clcb_stats_get_hostid@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_get_num(cl_engine*, clcb_stats_get_num)'    {cl_engine_set_clcb_stats_get_num@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_get_size(cl_engine*, clcb_stats_get_size)'    {cl_engine_set_clcb_stats_get_size@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_remove_sample(cl_engine*, clcb_stats_remove_sample)'    {cl_engine_set_clcb_stats_remove_sample@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_stats_submit(cl_engine*, clcb_stats_submit)'    {cl_engine_set_clcb_stats_submit@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_clcb_virus_found(cl_engine*, clcb_virus_found)'    {cl_engine_set_clcb_virus_found@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_set_num(cl_engine*, cl_engine_field, long long int)'    {cl_engine_set_num@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_set_stats_set_cbdata(cl_engine*, void*)'    {cl_engine_set_stats_set_cbdata@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_set_str(cl_engine*, cl_engine_field, const char*)'    {cl_engine_set_str@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_settings_apply(cl_engine*, const cl_settings*)'    {cl_engine_settings_apply@@CLAMAV_PUBLIC}
    [D] 'function cl_settings* cl_engine_settings_copy(const cl_engine*)'    {cl_engine_settings_copy@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_engine_settings_free(cl_settings*)'    {cl_engine_settings_free@@CLAMAV_PUBLIC}
    [D] 'function void cl_engine_stats_enable(cl_engine*)'    {cl_engine_stats_enable@@CLAMAV_PUBLIC}
    [D] 'function int cl_finish_hash(void*, void*)'    {cl_finish_hash@@CLAMAV_PUBLIC}
    [D] 'function void cl_fmap_close(cl_fmap_t*)'    {cl_fmap_close@@CLAMAV_PUBLIC}
    [D] 'function cl_fmap_t* cl_fmap_open_handle(void*, size_t, size_t, clcb_pread, int)'    {cl_fmap_open_handle@@CLAMAV_PUBLIC}
    [D] 'function cl_fmap_t* cl_fmap_open_memory(void*, size_t)'    {cl_fmap_open_memory@@CLAMAV_PUBLIC}
    [D] 'function void cl_hash_destroy(void*)'    {cl_hash_destroy@@CLAMAV_PUBLIC}
    [D] 'function void* cl_hash_init(const char*)'    {cl_hash_init@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_init(unsigned int)'    {cl_init@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_load(const char*, cl_engine*, unsigned int*, unsigned int)'    {cl_load@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_retdbdir()'    {cl_retdbdir@@CLAMAV_PUBLIC}
    [D] 'function unsigned int cl_retflevel()'    {cl_retflevel@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_retver()'    {cl_retver@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scandesc(int, const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*)'    {cl_scandesc@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scandesc_callback(int, const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*, void*)'    {cl_scandesc_callback@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scanfile(const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*)'    {cl_scanfile@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scanfile_callback(const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*, void*)'    {cl_scanfile_callback@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_scanmap_callback(cl_fmap_t*, const char*, const char**, unsigned long int*, const cl_engine*, cl_scan_options*, void*)'    {cl_scanmap_callback@@CLAMAV_PUBLIC}
    [D] 'function void cl_set_clcb_msg(clcb_msg)'    {cl_set_clcb_msg@@CLAMAV_PUBLIC}
    [D] 'function int cl_statchkdir(const cl_stat*)'    {cl_statchkdir@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_statfree(cl_stat*)'    {cl_statfree@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t cl_statinidir(const char*, cl_stat*)'    {cl_statinidir@@CLAMAV_PUBLIC}
    [D] 'function const char* cl_strerror(int)'    {cl_strerror@@CLAMAV_PUBLIC}
    [D] 'function int cl_update_hash(void*, void*, size_t)'    {cl_update_hash@@CLAMAV_PUBLIC}
    [D] 'function cl_error_t lsig_sub_matched(const cli_matcher*, cli_ac_data*, uint32_t, uint32_t, uint32_t, int)'    {lsig_sub_matched@@CLAMAV_PUBLIC}
```
This seems likely to break a number of users of libclamav.

Also, it seems like all of the rust symbols end up in the namespace, but I suppose that there is not much to be done about that at this point.

**Comment by opoplawski** [comment]
other apparent API change:
```
dbviewer.cpp:802:69: error: invalid conversion from 'unsigned int' to 'cl_error_t' [-fpermissive]
  802 |           printf("Database initialization error: %s\n", cl_strerror(ret));;
      |                                                                     ^~~
      |                                                                     |
      |                                                                     unsigned int
In file included from dbviewer.cpp:11:
/usr/include/clamav.h:1209:43: note:   initializing argument 1 of 'const char* cl_strerror(cl_error_t)'
 1209 | extern const char *cl_strerror(cl_error_t clerror);
      |                                ~~~~~~~~~~~^~~~~~~
```

**Comment by opoplawski** [comment]
Yeah, it appears that the whole CLAMAV_PUBLIC namespace has disappeared which seems quite problematic.  Is this intentional?  It seems like a big step backwards.

**Comment by micahsnyder** [comment]
> Yeah, it appears that the whole CLAMAV_PUBLIC namespace has disappeared which seems quite problematic. Is this intentional? It seems like a big step backwards.

It has been a long time since I put together the cmake build tooling.  From what I recall, I had trouble with finding a cross-platform way to replace `libclamav.map`.  But I don't recall what happened with trying to use the existing `libclamav.map`.  I vaguely remember searching for `.map` support with cmake and failing. 

The proposed code with `target_link_options( clamav PRIVATE "-Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/libclamav.map")` looks interesting. I'm looking at the PR more, now. 

> other apparent API change:
> 
> ```
> dbviewer.cpp:802:69: error: invalid conversion from 'unsigned int' to 'cl_error_t' [-fpermissive]
>   802 |           printf("Database initialization error: %s\n", cl_strerror(ret));;
>       |                                                                     ^~~
>       |                                                                     |
>       |                                                                     unsigned int
> In file included from dbviewer.cpp:11:
> /usr/include/clamav.h:1209:43: note:   initializing argument 1 of 'const char* cl_strerror(cl_error_t)'
>  1209 | extern const char *cl_strerror(cl_error_t clerror);
>       |                                ~~~~~~~~~~~^~~~~~~
> ```

This complaint is surprising to me.  From what I can tell, unless `-fshort-enums ` is used, `cl_error_t` (an enum) should be ABI compatible with `int` for C headers.   I see that the error is coming from a `.cpp` file.  But `clamav.h` does use 
```c
#ifdef __cplusplus
extern "C" {
#endif
```
... so that is still surprising to me.

**Comment by micahsnyder** [comment]
@opoplawski Can you please look at https://github.com/Cisco-Talos/clamav/pull/778 and tell me what you think?

**Comment by sthen** [comment]
Should cl_statchkdir() change from returning int to cl_error_t?

**Comment by micahsnyder** [comment]
> Should cl_statchkdir() change from returning int to cl_error_t?

Probably not. The possible return values right now are a bit of a mix...   If we did change it to `cl_error_t` then it would return `CL_VIRUS` instead of `1` if "some change occurred".  That feels a little weird to me.    

The documentation should at least be updated to say it may return some other value greater than 1 on error.

**Comment by sthen** [comment]
Ah. The context is that I ran into a package build failure in havp
(https://github.com/HaveSec/HAVP) with clamav 1.0.0; it tries to
call cl_strerror to decode failure returns from libclamav functions
(including from cl_statchkdir iff the return is not 0 or 1), and was
using a temporary int to store the return code.

To unbreak havp after updating clamav I patched away the cl_strerror
from the error message for now, which seems good enough, though it does
mean the relevant log is less informative if there is an error.
